metriq_gym/benchmarks/chsh.py

```diff
@@ -0,0 +1,266 @@
+            # indicating entanglement).
+            if exp_vals[idx] > 2:
+                good_edges.append(edge)
+    good_graph = rx.PyGraph(multigraph=False)
+    good_graph.add_nodes_from(list(range(coloring.num_nodes)))
+    for edge in good_edges:
+        good_graph.add_edge(*edge, 1)
+    return good_graph
+
+
+def largest_connected_size(good_graph: rx.PyGraph) -> int:
+    """Finds the size of the largest connected component in the CHSH subgraph.
+
+    Args:
+        good_graph: The graph of qubit pairs that violated CHSH inequality.
+
+    Returns:
+        The size of the largest connected component.
+    """
+    cc = rx.connected_components(good_graph)
+    largest_cc = cc[np.argmax([len(g) for g in cc])]
+    return len(largest_cc)
+
+
+class CHSH(Benchmark):
+    """Benchmark class for CHSH experiments."""
+
+    def dispatch_handler(self, device: QuantumDevice) -> CHSHData:
+        """Runs the benchmark and returns job metadata."""
+        shots = self.params.shots
+        topology_graph = None
+        coloring = None
+        trans_exp_sets = None
+
+        match device:
+            case QiskitBackend():
+                coloring = device_coloring(device)
+                exp_sets = generate_chsh_circuit_sets(coloring)
+                # The circuits are transpiled using a preset pass manager to optimize them for the device's constraints.
+                pm = generate_preset_pass_manager(1, device._backend)
+                trans_exp_sets = [pm.run(circ_set) for circ_set in exp_sets]
+
+            case IonQDevice():
+                topology_graph = nx.complete_graph(device.num_qubits)
+                raise ValueError("IonQ devices are not supported at this time.")
+
+            case BraketDevice():
+                topology_graph = device._device.topology_graph
+                raise ValueError("AWS devices are not supported at this time.")
+
+            case _:
+                raise ValueError(f"Unsupported device type: {type(device)}")
+
+        quantum_jobs: list[QuantumJob] = [
+            device.run(circ_set, shots=shots) for circ_set in trans_exp_sets
+        ]
+
+        provider_job_ids = [
+            job.id
+            for quantum_job_set in quantum_jobs
+            for job in (quantum_job_set if isinstance(quantum_job_set, list) else [quantum_job_set])
+        ]
+        return CHSHData(
+            provider_job_ids=provider_job_ids,
+            shots=shots,
+            num_qubits=device.num_qubits,
+            topology_graph=topology_graph,
+            coloring=coloring,
+        )
+
+    def poll_handler(self, job_data: BenchmarkData, result_data: list[ResultData]) -> None:
+        """Poll and calculate largest connected componenet."""
+        if not isinstance(job_data, CHSHData):
+            raise TypeError(f"Expected job_data to be of type {type(CHSHData)}")
+
+        good_graph = chsh_subgraph(job_data, result_data)
```

Comment:
Probably I am contradicting myself from what we discussed in the call, but what I would do here is extracting `.coloring` in the poll_handler, and passing only that to `chsh_subgraph`, so it's clearer that coloring is all you need there.

Reply:
Yeah, that makes more sense: c3a6d528eec713b94e5e34347b4b0b482b205660